Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1257,6 +1262,20 @@ public Task<Result<byte[]>> GetBlobAsync(OperationContext context, ContentHash h
             return GlobalStore.GetBlobAsync(context, hash);
         }
 
+        private async void OnContentLocationDatabaseInvalidationAsync(object sender, ContentLocationDatabase.DatabaseInvalidatedEventArgs e)
+        {
+            // If multiple threads fail at the same time (i.e. corruption error), this cheaply deduplicates the
+            // restores, avoids redundant logging.
+            using (_databaseInvalidationGate.AcquireSemaphore(0))
+            {
+                Tracer.Error(e.Context, $"Content location database has been invalidated. Forcing a restore from the last checkpoint. Error: {e.Failure.DescribeIncludingInnerFailures()}");
+
+                // There is nothing we can do from here if the checkpoint fails to restore. We will just wait until the
+                // next heartbeat.
+                await ProcessStateAsync(e.Context, inline: false, forceRestore: true).IgnoreFailure();

```

Comment:
>ProcessStateAsync [](start = 22, length = 17)

Somewhere we need to ensure that this isn't run concurrently with Heartbeat.

Reply:
Ok. I confused ProcessStateAsync with RestoreCheckpointAsync. RestoreCheckpointAsync makes sure that only one thread runs the operation, but there is no protection inside ProcessStateAsync.

Maybe we should just restart the timer here?

---
In reply to: [327875373](https://github.com/microsoft/BuildXL/pull/923#discussion_r327875373) [](ancestors = 327875373)

Reply:
End of the day we need to make sure that there is only one caller doing the Heartbeat, I have added a non-blocking semaphore in the heartbeat so as to avoid the undesired situations, and call directly the HeartbeatAsync function